File: alex/applications/PublicTransportInfoCS/data/expand_stops.pl
#!/usr/bin/env perl

use strict;
use warnings;
use utf8;
use Treex::Tool::Lexicon::Generation::CS;
use Treex::Tool::Lexicon::CS;
use Math::Cartesian::Product;
use autodie;
use Getopt::Long;
use Treex::Core::Common;
use Treex::Core::Scenario;
use Treex::Core::Document;

binmode(STDERR, ":utf8");

my $cases_list = '1,2,4';
my $merge_with = undef;
GetOptions( 'cases|c=s' => \$cases_list,
            'merge-with|merge|m=s' => \$merge_with );

if ( @ARGV != 2 ) {
    die("Usage: ./expand_stops.pl [--cases=1,2,4] stops.txt stops.expanded.txt\n");
}

open( my $in, '<:utf8', $ARGV[0] );
open( my $out, ">:utf8", $ARGV[1] );

my $generator = Treex::Tool::Lexicon::Generation::CS->new();
my $scenario_string = "Util::SetGlobal language=cs W2A::CS::Tokenize "
        . "W2A::CS::TagFeaturama lemmatize=1 W2A::CS::FixMorphoErrors";
my $scenario = Treex::Core::Scenario->new( { from_string => $scenario_string } );

log_set_error_level('WARN');

# Analyze one stop (input string) using Treex, return factored form|lemma|tag format as an array
sub analyze_stop {
    my ($source_text) = @_;
    my $document = Treex::Core::Document->new();
    my $bundle   = $document->create_bundle();
    my $zone     = $bundle->create_zone( 'cs', '' );
    $zone->set_sentence($source_text);

    $scenario->apply_to_documents($document);

    my @analysis = ();
    foreach my $bundle ( $document->get_bundles ) {
        my $zone = $bundle->get_zone( 'cs', '' );
        my @tokens = $zone->get_atree()->get_descendants( { ordered => 1 } );
        foreach my $token (@tokens){
            push @analysis, join('|', $token->form, $token->lemma, $token->tag);
        }
    }
    return \@analysis;    
}

# Given a factored form|lemma|tag array, return just forms
sub analyzed_to_plain {
    my @analyzed = @{ shift; };
    my $str = join(' ', map { $_ =~ s/\|.*//; $_ } @analyzed);
    $str =~ s/ ([\.,])/$1/g;
    return $str;
}

#
# MAIN
#

my %merge = ();

# Load stop names and forms to be merged with the input
if ($merge_with){
    open( my $merge_in, '<:utf8', $merge_with );
    while ( my $line = <$merge_in> ){
        chomp $line;
        my @variants = split /;/, $line;
        my $main_stop_name = $variants[0];
        # merge inflection forms for all repeating lines
        if (defined($merge{$main_stop_name})){
            my %var_hash = map { $_ => 1 } @{$merge{$main_stop_name}};
            @variants = grep { not defined($var_hash{$_}) } @variants;
            push @{$merge{$main_stop_name}}, @variants;
        }
        # first occurence of a line (normal case): just remember all
        else {
            $merge{$main_stop_name} = \@variants;
        }
    }
    close($merge_in);
}

# Read each stop and try to inflect it in different cases
my $line_ctr = 0;
while ( my $line = <$in> ) {
    
    chomp $line;
    # skip commented or empty lines
    if ($line =~ /^#/ or $line eq ''){
        next;
    }

    $line_ctr++;
    if ($line_ctr % 10 == 0){
        print STDERR '.';
    }
    my @cases;  # store all inflected forms here

    # analyze all variant names
    my @variants = map { analyze_stop($_); } split(/;/, $line);
    my $main_stop_name = analyzed_to_plain($variants[0]);
    
    # filter variants if already inflected
    if (defined($merge{$main_stop_name})){
        push @cases, @{$merge{$main_stop_name}};
        my %merge_hash = map { $_ => 1 } @{$merge{$main_stop_name}};
        @variants = grep { my $plainvar = analyzed_to_plain($_); not defined($merge_hash{$plainvar}); } @variants;
    }

    # process all possible variant names of the given stop/city
    foreach my $variant (@variants){
        
        my @words = @$variant;
    
        # try all needed cases
        foreach my $case (split /[, ]/, $cases_list) {
    
            my @forms;
            my $prevtag = '';
    
            # try inflecting each word in nominative not following a noun in nominative
            # (only if current case is not nominative)
            foreach my $word (@words) {
    
                my ( $form, $lemma, $tag ) = split /\|/, $word;
    
                if ( $tag =~ /^....1/ and $prevtag !~ /^NN..1/ and $case ne '1' ) {
    
                    my $newtag = $tag;
                    $newtag =~ s/^(....)1/$1$case/;
                    $newtag =~ s/.$/./;
    
                    # -ice: test add both sg. and pl. versions
                    if ( $form =~ /ice$/ and $tag =~ /^...S/ ) {
                        $newtag =~ s/^(...)./${1}[SP]/;
                    }
    
                    # retrieve the inflection form(s)
                    my @newforms = $generator->forms_of_lemma( $lemma, { tag_regex => $newtag } );
    
                    if (@newforms) {
                        # preserve capitalization
                        @newforms = map { substr( $form, 0, 1 ) . substr( $_->get_form(), 1 ) } @newforms;
                    }
    
                    # fallback to uninflected (if form not found in the dictionary)
                    else {
                        print STDERR "\nCannot inflect: $form, $lemma, $tag\n";
                        @newforms = ($form);
                    }
    
                    push @forms, \@newforms;
                }
                else {
                    push @forms, [$form];
                }
                $prevtag = $tag;
            }
    
            # merge all variants of the same case (+ avoid repeating)
            cartesian {
                my $forms_str = join( ' ', @_ );
                $forms_str =~ s/ ([\.,])/$1/g;  # avoid spaces before comma or period
                push @cases, $forms_str if ( !grep { $_ eq $forms_str } @cases );
            }
            @forms;
        }
    }
    print {$out} join( ";", @cases ), "\n";
}

print STDERR "\n";

__END__

=encoding utf8

=head1 NAME

expand_stops.pl -- PID stop inflection using Treex (http://ufal.mff.cuni.cz/treex)

=head1 DESCRIPTION

Inflecting PID stops in all cases using Treex (morphological analyzer,
morphological generator & tagger for Czech).

=head1 USAGE

./gen_stops.pl [--cases=1,2,4] [--merge=earlier-stops.expanded.txt] stops.txt stops.expanded.txt

Cases default to 1,2,4 (nominative, genitive, accusative). Use --merge for incremental merging
(all earlier forms will be preserved, including additional hand-written ones).

=head1 AUTHOR

Ondřej Dušek

=head1 COPYRIGHT AND LICENSE

Copyright © 2013 by Institute of Formal and Applied Linguistics, Charles University in Prague

Distributed under the Apache 2.0 license

File: alex/applications/PublicTransportInfoCS/data/expand_stops.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Expanding Czech stop/city/train names with inflection forms in all given morphological cases.

Usage:

./expand_stops.py [-c 1,2,4] [-n] [-l] [-p] [-m previous.expanded.txt] stops.txt [stops-2.txt ...] stops.expanded.txt

-c = Required morphological cases (possible values: numbers 1-7, comma-separated).
        Defaults to 1,2,4 (nominative, genitive, accusative).
-m = Add previously expanded stops. Only stop name variants NOT found in the previously
        expanded file are expanded to speed up the process and preserve hand-written forms.
-p = Strip punctuation off surface forms
-l = Lowercase surface forms
-n = Inflecting personal names (use this for train names, do not use for stop/city names)
"""
from __future__ import unicode_literals
if __name__ == '__main__':
    import autopath

import sys
import re
from collections import defaultdict
import itertools
import codecs

from alex.utils.config import online_update
from alex.utils.various import remove_dups_stable
from alex.applications.PublicTransportInfoCS.cs_morpho import Analyzer, Generator 
from getopt import getopt


class ExpandStops(object):
    """This handles inflecting stop names into all desired cases in Czech."""

    def __init__(self, cases_list, strip_punct, lowercase_forms, personal_names):
        """Initialize the expander object, initialize the morphological analyzer and generator.

        @param cases_list: List of cases (given as strings) to be used for generation \
                (Czech numbers 1-7 are used)
        @param strip_punct: Strip all punctuation ?
        @param lowercase_forms: Lowercase all forms on the output?
        @param personal_names: Are we inflecting personal names?
        """
        self.stops = defaultdict(list)
        self.cases_list = cases_list
        self.personal_names = personal_names
        # initialize postprocessing
        postprocess_func = ((lambda text: re.sub(r' ([\.,])', r'\1', text))
                            if not strip_punct
                            else (lambda text: re.sub(r' [\.,\-–\(\)\{\}\[\];\\\/+&](?: [\.,\-–\(\)\{\}\[\];])*( |$)', r'\1', text)))
        if lowercase_forms:
            lc_func = lambda text: postprocess_func(text).lower()
            self.__postprocess_func = lc_func
        else:
            self.__postprocess_func = postprocess_func
        # initialize morphology
        analyzer_model = online_update('applications/PublicTransportInfoCS/data/czech.tagger')
        generator_model = online_update('applications/PublicTransportInfoCS/data/czech.dict')
        self.__analyzer = Analyzer(analyzer_model)
        self.__generator = Generator(generator_model)

    def save(self, fname):
        """Save all stops currently held in memory to a file."""
        with codecs.open(fname, 'w', 'UTF-8') as f_out:
            for stop_name in sorted(self.stops.keys()):
                f_out.write(stop_name + "\t")
                f_out.write('; '.join(self.stops[stop_name]))
                f_out.write("\n")

    def parse_line(self, line):
        """Load one line from the input file (tab-separated main form or
        implicit main form supported)."""
        if '\t' not in line:
            stop = None
            variants = line
        else:
            stop, variants = line.split('\t')
        variants = [var.strip() for var in variants.split(';')]
        if stop is None:
            stop = variants[0]
        return stop, variants

    def load_file(self, fname):
        """Just load a list of stops from a file and store it in memory."""
        with codecs.open(fname, 'r', 'UTF-8') as f_in:
            for line in f_in:
                if line.startswith('#'):  # skip comments
                    continue
                stop, variants = self.parse_line(line)
                self.stops[stop] = list(remove_dups_stable(variants + self.stops[stop]))

    def expand_file(self, fname):
        """Load a list of stops from a file and expand it."""
        with codecs.open(fname, 'r', 'UTF-8') as f_in:
            ctr = 0
            for line in f_in:
                if line.startswith('#'):  # skip comments
                    continue
                # load variant names for a stop
                stop, variants = self.parse_line(line)
                # skip those that needn't be inflected any more
                to_inflect = [var for var in variants if not var in self.stops[stop]]
                # inflect the rest
                for variant in to_inflect:
                    words = self.__analyzer.analyze(variant)
                    # in all required cases
                    for case in self.cases_list:
                        forms = self.__generator.inflect(words, case, self.personal_names)
                        # use all possible combinations if there are more variants for this case
                        inflected = map(self.__postprocess_func,
                                        remove_dups_stable([' '.join(var)
                                                            for var in itertools.product(*forms)]))
                        self.stops[stop] = list(remove_dups_stable(self.stops[stop] + inflected))
                ctr += 1
                if ctr % 1000 == 0:
                    print >> sys.stderr, '.',
        print >> sys.stderr


def main():
    cases = ['1', '2', '4']
    merge_file = None
    strip_punct = False
    lowercase_forms = False
    personal_names = False
    opts, files = getopt(sys.argv[1:], 'c:m:pln')
    for opt, arg in opts:
        if opt == '-c':
            cases = re.split('[, ]+', arg)
        elif opt == '-m':
            merge_file = arg
        elif opt == '-p':
            strip_punct = True
        elif opt == '-l':
            lowercase_forms = True
        elif opt == '-n':
            personal_names = True
    if len(files) < 2:
        sys.exit(__doc__)
    in_files = files[:-1]
    out_file = files[-1]
    es = ExpandStops(cases, strip_punct, lowercase_forms, personal_names)
    if merge_file:
        es.load_file(merge_file)
    for in_file in in_files:
        es.expand_file(in_file)
    es.save(out_file)


if __name__ == '__main__':
    main()
